{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evaluate SLAV-seq coverage of rmsk, and knrgl\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subprocess import Popen, PIPE, DEVNULL\n",
    "from tempfile import NamedTemporaryFile\n",
    "from io import StringIO\n",
    "from collections import defaultdict\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import pysam\n",
    "\n",
    "from tqdm import tqdm\n",
    "\n",
    "from scripts.get_labels import read_knrgl\n",
    "from myutils.rmsk import read_rmsk\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find knrgl bed and bulk BAMs for each individual\n",
    "individuals = pd.read_csv(\n",
    "    \"/iblm/logglun02/mcuoco/workflows/sz_slavseq/config/bulk_donors.tsv\", sep=\"\\t\"\n",
    ")[\"donor_id\"].values.astype(str)\n",
    "indv_data = {i: {\"knrgl\": None, \"bulk\": None} for i in individuals}\n",
    "\n",
    "for bulk in Path(\"/iblm/netapp/data4/mcuoco/sz_slavseq/results/align/\").rglob(\n",
    "    \"*/gDNA_usd*.sorted.bam\"\n",
    "):\n",
    "    if bulk.parts[-2] in individuals:\n",
    "        indv_data[bulk.parts[-2]][\"bulk\"] = str(bulk)\n",
    "\n",
    "for knrgl in Path(\"/iblm/netapp/data4/mcuoco/sz_slavseq/resources/\").rglob(\n",
    "    \"*_insertions.bed\"\n",
    "):\n",
    "    if knrgl.name.split(\"_\")[0] in individuals:\n",
    "        indv_data[knrgl.name.split(\"_\")[0]][\"knrgl\"] = str(knrgl)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate coverage of RMSK and KNRGL per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define helper functions\n",
    "def bedtools_intersect(bam, bed, view_args=\"\"):\n",
    "    cmd = f\"samtools view -b {bam} {view_args} | bedtools intersect -abam stdin -b {bed} -bed -wa -wb\"\n",
    "    p = Popen(cmd, shell=True, stdout=PIPE, stderr=DEVNULL)\n",
    "    names = [\n",
    "        \"Chromosome\",\n",
    "        \"Start\",\n",
    "        \"End\",\n",
    "        \"read_id\",\n",
    "        \"mapq\",\n",
    "        \"Strand\",\n",
    "        \"intersect_start\",\n",
    "        \"intersect_end\",\n",
    "        \"x\",\n",
    "        \"y\",\n",
    "        \"Flag\",\n",
    "        \"z\",\n",
    "        \"ChromosomeB\",\n",
    "        \"StartB\",\n",
    "        \"EndB\",\n",
    "        \"NameB\",\n",
    "        \"ScoreB\",\n",
    "        \"StrandB\",\n",
    "    ]\n",
    "\n",
    "    with StringIO(p.stdout.read().decode()) as bed:\n",
    "        df = pd.read_csv(bed, sep=\"\\t\", header=None, names=names)\n",
    "\n",
    "    df.drop([\"x\", \"y\", \"z\"], axis=1, inplace=True)\n",
    "    df[\"Flag\"] = df[\"Flag\"].str.rstrip(\",\").astype(int)\n",
    "    return df\n",
    "\n",
    "\n",
    "def bedtools_inverse_intersect(bam, bed, view_args=\"\"):\n",
    "    cmd = f\"samtools view -b {bam} {view_args} | bedtools intersect -abam stdin -bed -v -wa -b\"\n",
    "\n",
    "    if isinstance(bed, list):\n",
    "        for b in bed:\n",
    "            cmd += f\" {b}\"\n",
    "    else:\n",
    "        cmd += f\" {bed}\"\n",
    "\n",
    "    p = Popen(cmd, shell=True, stdout=PIPE, stderr=DEVNULL)\n",
    "    names = [\n",
    "        \"Chromosome\",\n",
    "        \"Start\",\n",
    "        \"End\",\n",
    "        \"read_id\",\n",
    "        \"mapq\",\n",
    "        \"Strand\",\n",
    "        \"intersect_start\",\n",
    "        \"intersect_end\",\n",
    "        \"x\",\n",
    "        \"y\",\n",
    "        \"Flag\",\n",
    "        \"z\",\n",
    "    ]\n",
    "    with StringIO(p.stdout.read().decode()) as bed:\n",
    "        df = pd.read_csv(bed, sep=\"\\t\", header=None, names=names, on_bad_lines=\"skip\")\n",
    "\n",
    "    df.drop([\"x\", \"y\", \"z\"], axis=1, inplace=True)\n",
    "    df[\"Flag\"] = df[\"Flag\"].str.rstrip(\",\").astype(int)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in repeatmasker output\n",
    "rmsk = read_rmsk(\"/iblm/netapp/data4/mcuoco/sz_slavseq/resources/hs38d1.fa.out\")\n",
    "\n",
    "rep_names = [\n",
    "    \"L1HS_3end\",\n",
    "    \"L1PA2_3end\",\n",
    "    \"L1PA3_3end\",\n",
    "    \"L1PA4_3end\",\n",
    "    \"L1PA5_3end\",\n",
    "    \"L1PA6_3end\",\n",
    "]\n",
    "\n",
    "rmsk = rmsk.loc[(rmsk[\"repName\"].isin(rep_names)) & (rmsk[\"repEnd\"] > 860), :]\n",
    "rmsk = rmsk.apply(\n",
    "    lambda x: x\n",
    "    if (x[\"strand\"] == \"+\" and x[\"repStart\"] < 765)\n",
    "    or (x[\"strand\"] == \"-\" and x[\"repLeft\"] < 765)\n",
    "    else None,\n",
    "    axis=1,\n",
    ").dropna()\n",
    "\n",
    "rmsk[\"genoStart\"] = rmsk.apply(\n",
    "    lambda x: x[\"genoStart\"] - 1000 if x[\"strand\"] == \"-\" else x[\"genoStart\"], axis=1\n",
    ")\n",
    "rmsk[\"genoEnd\"] = rmsk.apply(\n",
    "    lambda x: x[\"genoEnd\"] + 1000 if x[\"strand\"] == \"+\" else x[\"genoEnd\"], axis=1\n",
    ")\n",
    "\n",
    "rmsk = rmsk.rename(\n",
    "    columns={\n",
    "        \"genoName\": \"Chromosome\",\n",
    "        \"genoStart\": \"Start\",\n",
    "        \"genoEnd\": \"End\",\n",
    "        \"strand\": \"Strand\",\n",
    "    }\n",
    ").loc[:, [\"Chromosome\", \"Start\", \"End\", \"Strand\"]]\n",
    "rmsk = rmsk.loc[(rmsk.Start >= 0) & (rmsk.End >= 0), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define read groups to search\n",
    "read_groups = {\n",
    "    \"all\": \"\",\n",
    "    \"read1\": \"-f 64\",\n",
    "    \"read2\": \"-f 128\",\n",
    "    \"primary read1\": \"-F 256 -f 64\",\n",
    "    \"primary read2\": \"-F 256 -f 128\",\n",
    "    \"proper pair\": \"-f 2\",\n",
    "    \"primary proper pair\": \"-F 256 -f 2\",\n",
    "    \"not proper pair primary read1\": \"-F 258 -f 128\",\n",
    "    \"not proper pair primary read2\": \"-F 258 -f 64\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# collect knrgl and rmsk coverage info for each donor that has 90x WGS (repeat once 30x WGS is available)\n",
    "\n",
    "rmsk_tmp = NamedTemporaryFile()\n",
    "pr.PyRanges(rmsk).to_bed(rmsk_tmp.name)\n",
    "\n",
    "res = []\n",
    "# ~30 min per iteration\n",
    "# for each donor\n",
    "for indv, data in tqdm(indv_data.items(), total=len(indv_data)):\n",
    "    bam = data[\"bulk\"]\n",
    "    knrgl = read_knrgl(data[\"knrgl\"])\n",
    "\n",
    "    knrgl_tmp = NamedTemporaryFile()\n",
    "    pr.PyRanges(knrgl).to_bed(knrgl_tmp.name)\n",
    "\n",
    "    # for each read group\n",
    "    for i, (k, v) in enumerate(read_groups.items()):\n",
    "\n",
    "        # total reads\n",
    "        if k == \"all\":\n",
    "            total_reads = int(pysam.view(bam, \"-c\").rstrip(\"\\n\"))\n",
    "        else:\n",
    "            total_reads = int(pysam.view(bam, v, \"-c\").rstrip(\"\\n\"))\n",
    "\n",
    "        # reads outside rmsk, knrgl, and blast\n",
    "        outside_reads = bedtools_inverse_intersect(\n",
    "            bam, [rmsk_tmp.name, knrgl_tmp.name], view_args=v\n",
    "        )\n",
    "\n",
    "        # reads overlapping rmsk and knrgl\n",
    "        knrgl_reads = bedtools_intersect(bam, knrgl_tmp.name, view_args=v)\n",
    "        knrgl_reads[\"label\"] = \"KNRGL\"\n",
    "        rmsk_reads = bedtools_intersect(bam, rmsk_tmp.name, view_args=v)\n",
    "        rmsk_reads[\"label\"] = \"RMSK\"\n",
    "\n",
    "        # get number of reads per region\n",
    "        df = (\n",
    "            pd.concat([knrgl_reads, rmsk_reads])\n",
    "            .groupby([\"ChromosomeB\", \"StartB\", \"EndB\", \"label\"])\n",
    "            .size()\n",
    "            .reset_index(name=\"reads\")\n",
    "        )\n",
    "\n",
    "        df[\"read_group\"] = k\n",
    "        df[\"indv\"] = indv\n",
    "        df[\"total_reads\"] = total_reads\n",
    "        df[\"outside_reads\"] = len(outside_reads)\n",
    "\n",
    "        res.append(df)\n",
    "\n",
    "    knrgl_tmp.close()\n",
    "\n",
    "rmsk_tmp.close()\n",
    "res = pd.concat(res)\n",
    "\n",
    "res.to_csv(\"indv_coverage.tsv\", sep=\"\\t\", index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analyze results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = pd.read_csv(\"indv_coverage.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create facetgrid\n",
    "g = sns.FacetGrid(\n",
    "    res[res[\"read_group\"] == \"all\"],\n",
    "    col=\"label\",\n",
    "    hue=\"indv\",\n",
    "    sharex=False,\n",
    "    sharey=True,\n",
    "    height=4.5,\n",
    "    aspect=1,\n",
    "    gridspec_kws={\"wspace\": 0.5},\n",
    ")\n",
    "g.map_dataframe(sns.ecdfplot, x=\"reads\", stat=\"proportion\").set(xscale=\"log\")\n",
    "g.add_legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create facetgrid\n",
    "g = sns.FacetGrid(\n",
    "    res[res[\"label\"] == \"KNRGL\"],\n",
    "    col=\"read_group\",\n",
    "    hue=\"indv\",\n",
    "    sharex=False,\n",
    "    sharey=True,\n",
    "    height=4.5,\n",
    "    aspect=1,\n",
    "    col_wrap=3,\n",
    "    gridspec_kws={\"wspace\": 0.5},\n",
    ")\n",
    "g.map_dataframe(sns.ecdfplot, x=\"reads\", stat=\"proportion\").set(xscale=\"log\")\n",
    "g.add_legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create facetgrid\n",
    "g = sns.FacetGrid(\n",
    "    res[res[\"label\"] == \"RMSK\"],\n",
    "    col=\"read_group\",\n",
    "    hue=\"indv\",\n",
    "    sharex=False,\n",
    "    sharey=True,\n",
    "    height=4.5,\n",
    "    aspect=1,\n",
    "    col_wrap=3,\n",
    "    gridspec_kws={\"wspace\": 0.5},\n",
    ")\n",
    "g.map_dataframe(sns.ecdfplot, x=\"reads\", stat=\"proportion\").set(xscale=\"log\")\n",
    "g.add_legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# wrangle for plotting\n",
    "plot_df = (\n",
    "    res.drop([\"ChromosomeB\", \"StartB\", \"EndB\"], axis=1)\n",
    "    .groupby([\"read_group\", \"indv\", \"total_reads\", \"outside_reads\", \"label\"])\n",
    "    .count()\n",
    "    .reset_index()\n",
    "    .rename(columns={\"reads\": \"covered\"})\n",
    "    .pivot(\n",
    "        index=[\"read_group\", \"indv\", \"total_reads\", \"outside_reads\"],\n",
    "        columns=\"label\",\n",
    "        values=\"covered\",\n",
    "    )\n",
    "    .reset_index()\n",
    "    .rename(columns={\"KNRGL\": \"knrgl_covered\", \"RMSK\": \"rmsk_covered\"})\n",
    ")\n",
    "\n",
    "plot_df[\"total_knrgl\"] = pd.Series([1] * len(plot_df))\n",
    "plot_df[\"total_rmsk\"] = pd.Series([1] * len(plot_df))\n",
    "\n",
    "plot_df.indv = plot_df.indv.astype(str)\n",
    "for indv, data in indv_data.items():\n",
    "    knrgl = read_knrgl(data[\"knrgl\"])\n",
    "    plot_df.loc[(plot_df.indv == indv), \"total_knrgl\"] = len(knrgl)\n",
    "    plot_df.loc[(plot_df.indv == indv), \"total_rmsk\"] = len(rmsk)\n",
    "\n",
    "plot_df[\"frac knrgl covered\"] = plot_df.knrgl_covered / plot_df.total_knrgl\n",
    "plot_df[\"frac rmsk covered\"] = plot_df.rmsk_covered / plot_df.total_rmsk\n",
    "\n",
    "plot_df[\"inside_reads\"] = plot_df.total_reads - plot_df.outside_reads\n",
    "plot_df[\"frac alignments covering L1\"] = plot_df.inside_reads / plot_df.total_reads\n",
    "\n",
    "plot_df.rename(columns={\"total_reads\": \"total alignments\"}, inplace=True)\n",
    "\n",
    "plot_df = (\n",
    "    plot_df[\n",
    "        [\n",
    "            \"read_group\",\n",
    "            \"indv\",\n",
    "            \"total alignments\",\n",
    "            \"frac knrgl covered\",\n",
    "            \"frac rmsk covered\",\n",
    "            \"frac alignments covering L1\",\n",
    "        ]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .melt(\n",
    "        id_vars=[\"read_group\", \"indv\"],\n",
    "        value_vars=[\n",
    "            \"total alignments\",\n",
    "            \"frac alignments covering L1\",\n",
    "            \"frac knrgl covered\",\n",
    "            \"frac rmsk covered\",\n",
    "        ],\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.catplot(\n",
    "    plot_df[plot_df[\"read_group\"] == \"all\"],\n",
    "    y=\"indv\",\n",
    "    x=\"value\",\n",
    "    col=\"label\",\n",
    "    kind=\"bar\",\n",
    "    sharey=False,\n",
    "    sharex=False,\n",
    "    height=4,\n",
    "    aspect=1,\n",
    "    legend=True,\n",
    "    col_wrap=2,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.catplot(\n",
    "    plot_df,\n",
    "    x=\"indv\",\n",
    "    y=\"value\",\n",
    "    col=\"read_group\",\n",
    "    row=\"label\",\n",
    "    kind=\"bar\",\n",
    "    sharey=\"row\",\n",
    "    sharex=False,\n",
    "    height=4,\n",
    "    aspect=1,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "shell",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
