{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import warnings\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "cols = sns.color_palette(\"tab10\", n_colors=2)\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as colors\n",
    "import pyranges as pr\n",
    "from collections import defaultdict\n",
    "from tqdm.notebook import tqdm\n",
    "from sklearn.metrics import roc_curve, roc_auc_score, RocCurveDisplay\n",
    "from pysam import AlignmentFile\n",
    "from tempfile import NamedTemporaryFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_megane_bed(bed_fn):\n",
    "    df = pd.read_csv(\n",
    "        bed_fn,\n",
    "        sep=\"\\t\",\n",
    "        header=None,\n",
    "        names=[\n",
    "            \"Chromosome\",\n",
    "            \"Start\",\n",
    "            \"End\",\n",
    "            \"family\",\n",
    "            \"left_bp\",\n",
    "            \"right_bp\",\n",
    "            \"confidence\",\n",
    "            \"length\",\n",
    "            \"subfamily\",\n",
    "            \"transduction\",\n",
    "            \"filter\",\n",
    "        ],\n",
    "        usecols=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],\n",
    "    )\n",
    "    df = df[df[\"filter\"].str.contains(\"PASS;H\")]\n",
    "    df = df[df[\"subfamily\"].str.contains(\"L1HS\")]\n",
    "    df = df[df[\"confidence\"].str.contains(\"high;PASS\")]\n",
    "    df[\"donor_AC\"] = df[\"filter\"].str.split(\";\").str[0].astype(int)\n",
    "    return df\n",
    "\n",
    "\n",
    "# read meta data\n",
    "samples = pd.read_csv(\n",
    "    snakemake.config[\"samples\"],\n",
    "    sep=\"\\t\",\n",
    "    dtype={\"sample_id\": str, \"tissue_id\": str, \"donor_id\": str},\n",
    ")\n",
    "meta = pd.read_csv(snakemake.config[\"donors\"], sep=\"\\t\", dtype={\"donor_id\": str})\n",
    "samples = samples.merge(meta, on=[\"donor_id\"]).set_index(\n",
    "    [\"donor_id\", \"sample_id\"], drop=False\n",
    ")\n",
    "\n",
    "# read coverage\n",
    "cov = {Path(c).name.rstrip(\".coverage.bed\"): c for c in snakemake.input.cov}\n",
    "samples[\"coverage\"] = samples[\"sample_id\"].map(cov)\n",
    "\n",
    "with open(\"resources/bad_cells.txt\", \"r\") as f:\n",
    "    bad_cells = [line.strip() for line in f.readlines()]\n",
    "\n",
    "samples = samples[~samples[\"sample_id\"].isin(bad_cells)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cov = []\n",
    "for r in tqdm(samples.itertuples(), total=len(samples)):\n",
    "    mdf = read_megane_bed(r.megane_30x)[[\"Chromosome\", \"Start\", \"End\", \"donor_AC\"]]\n",
    "    mdf = pr.PyRanges(mdf).extend(100).df\n",
    "    cdf = pd.read_csv(\n",
    "        r.coverage,\n",
    "        sep=\"\\t\",\n",
    "        header=None,\n",
    "        names=[\"Chromosome\", \"Start\", \"End\", \"reads\", \"rpm\"],\n",
    "    )\n",
    "    cdf[\"cell_id\"] = r.sample_id\n",
    "    cdf[\"donor_id\"] = r.donor_id\n",
    "    cdf[\"locus\"] = (\n",
    "        cdf[\"Chromosome\"].astype(str)\n",
    "        + \":\"\n",
    "        + cdf[\"Start\"].astype(str)\n",
    "        + \"-\"\n",
    "        + cdf[\"End\"].astype(str)\n",
    "    )\n",
    "    cdf = cdf.join(\n",
    "        mdf.set_index([\"Chromosome\", \"Start\", \"End\"]), on=[\"Chromosome\", \"Start\", \"End\"]\n",
    "    )\n",
    "    cdf[\"donor_AC\"] = cdf[\"donor_AC\"].fillna(0)\n",
    "    cov.append(cdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cov = pd.concat(cov).reset_index(drop=True)\n",
    "cov[\"label\"] = cov[\"donor_AC\"].apply(lambda x: True if x > 0 else False)\n",
    "bulk_cov = cov[\n",
    "    cov[\"cell_id\"].str.contains(\"gDNA\") | cov[\"cell_id\"].str.contains(\"Bulk\")\n",
    "].copy()\n",
    "af = bulk_cov.groupby(\"locus\")[\"donor_AC\"].sum()\n",
    "bulk_cov[\"AC\"] = bulk_cov[\"locus\"].apply(lambda x: af[x])\n",
    "cell_cov = cov[\n",
    "    ~cov[\"cell_id\"].str.contains(\"gDNA\") & ~cov[\"cell_id\"].str.contains(\"Bulk\")\n",
    "].copy()\n",
    "cell_cov[\"AC\"] = cell_cov[\"locus\"].apply(lambda x: af[x])\n",
    "cell_cov[\"tissue\"] = cell_cov[\"cell_id\"].apply(\n",
    "    lambda x: \"HIP\" if \"ush\" in x.lower() else \"DLPFC\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pairwise concordance of variant calls with SLAVseq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bulk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make df of zeros with donor_id as index and columns\n",
    "donors = meta.sort_values(\"race\")[meta.donor_id != \"CommonBrain\"][\"donor_id\"]\n",
    "df = pd.DataFrame(0, index=donors, columns=donors)\n",
    "\n",
    "for d1 in donors:\n",
    "    d1_knrgls = bulk_cov.query(\"donor_id == @d1 and donor_AC > 0\")[\"locus\"].unique()\n",
    "    for d2 in donors:\n",
    "        d2_cov = len(\n",
    "            bulk_cov[\n",
    "                bulk_cov[\"donor_id\"].eq(d2)\n",
    "                & bulk_cov[\"locus\"].isin(d1_knrgls)\n",
    "                & bulk_cov[\"rpm\"].ge(10)\n",
    "            ]\n",
    "        )\n",
    "        df.loc[d1, d2] = d2_cov / len(d1_knrgls)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "races = meta.set_index(\"donor_id\").loc[donors][\"race\"]\n",
    "race_cols = {\"CAUC\": cols[0], \"AA\": cols[1]}\n",
    "row_colors = races.map(race_cols)\n",
    "\n",
    "plt.figure(figsize=(10, 10))\n",
    "plt.title(\"% 30x WGS LINE-1 calls covered by donor X bulk SLAVseq\")\n",
    "ax = sns.heatmap(df, square=True)\n",
    "ax.set_xlabel(\"Donor Bulk SLAVseq\")\n",
    "ax.set_ylabel(\"Donor WGS LINE-1 calls\")\n",
    "ax.tick_params(\n",
    "    axis=\"y\", which=\"major\", pad=12, length=0\n",
    ")  # extra padding to leave room for the row colors\n",
    "ax.tick_params(\n",
    "    axis=\"x\", which=\"major\", pad=12, length=0\n",
    ")  # extra padding to leave room for the row colors\n",
    "\n",
    "for i, color in enumerate(row_colors):\n",
    "    ax.add_patch(\n",
    "        plt.Rectangle(\n",
    "            xy=(-0.02, i),\n",
    "            width=0.02,\n",
    "            height=1,\n",
    "            color=color,\n",
    "            lw=0,\n",
    "            transform=ax.get_yaxis_transform(),\n",
    "            clip_on=False,\n",
    "        )\n",
    "    )\n",
    "    ax.add_patch(\n",
    "        plt.Rectangle(\n",
    "            xy=(i, -0.02),\n",
    "            width=1,\n",
    "            height=0.02,\n",
    "            color=color,\n",
    "            lw=0,\n",
    "            transform=ax.get_xaxis_transform(),\n",
    "            clip_on=False,\n",
    "        )\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Single-cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reshape in 2d\n",
    "cell_cov_2d = cell_cov.pivot(index=\"locus\", columns=\"cell_id\", values=\"reads\")\n",
    "\n",
    "# get donors and cells\n",
    "donors = meta.sort_values(\"race\")[meta.donor_id != \"CommonBrain\"][\"donor_id\"]\n",
    "cells = (\n",
    "    cell_cov[cell_cov.donor_id != \"CommonBrain\"]\n",
    "    .set_index(\"donor_id\")\n",
    "    .sort_values(\"tissue\")\n",
    "    .loc[donors][\"cell_id\"]\n",
    "    .unique()\n",
    ")\n",
    "\n",
    "# make df of zeros with donor_id as index and columns\n",
    "df = pd.DataFrame(0, index=donors, columns=cells)\n",
    "with tqdm(total=len(donors) * len(cells)) as pbar:\n",
    "    for d in donors:\n",
    "        d_knrgls = bulk_cov[(bulk_cov.donor_id == d) & (bulk_cov.donor_AC > 0)][\n",
    "            \"locus\"\n",
    "        ].unique()\n",
    "        for c in cells:\n",
    "            c_cov = sum(cell_cov_2d.loc[d_knrgls, c] >= 10)\n",
    "            df.loc[d, c] = c_cov / len(d_knrgls)\n",
    "            pbar.update(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_cov"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = sns.color_palette(\"tab10\", n_colors=4)\n",
    "\n",
    "tissues = (\n",
    "    cell_cov.set_index(\"donor_id\", drop=False)\n",
    "    .loc[donors][[\"tissue\", \"cell_id\"]]\n",
    "    .drop_duplicates()[\"tissue\"]\n",
    ")\n",
    "tissue_cols = {\"HIP\": cols[3], \"DLPFC\": cols[2]}\n",
    "col_colors = tissues.map(tissue_cols)\n",
    "\n",
    "races = meta.set_index(\"donor_id\").loc[donors][\"race\"]\n",
    "race_cols = {\"CAUC\": cols[0], \"AA\": cols[1]}\n",
    "row_colors = races.map(race_cols)\n",
    "\n",
    "group_sizes = (\n",
    "    cell_cov[[\"tissue\", \"cell_id\", \"donor_id\"]]\n",
    "    .drop_duplicates()\n",
    "    .groupby([\"donor_id\", \"tissue\"])\n",
    "    .size()\n",
    "    .to_frame(name=\"size\")\n",
    "    .loc[donors]\n",
    ")\n",
    "group_sizes[\"cum\"] = group_sizes[\"size\"].cumsum()\n",
    "\n",
    "positions, labels = [], []\n",
    "for i, s in group_sizes.iterrows():\n",
    "    positions.append(s.cum - (s[\"size\"] / 2))\n",
    "    labels.append(i)\n",
    "\n",
    "plt.figure(figsize=(100, 10))\n",
    "plt.title(\"% 30x WGS LINE-1 calls covered by >=10 reads in single-cell SLAVseq\")\n",
    "ax = sns.heatmap(df)\n",
    "ax.set_xlabel(\"Single-cell SLAVseq\")\n",
    "ax.set_ylabel(\"Donor WGS LINE-1 calls\")\n",
    "ax.set_xticks(positions)\n",
    "ax.set_xticklabels(labels)\n",
    "\n",
    "ax.tick_params(\n",
    "    axis=\"y\", which=\"major\", pad=24, length=0\n",
    ")  # extra padding to leave room for the row colors\n",
    "for i, color in enumerate(row_colors):\n",
    "    ax.add_patch(\n",
    "        plt.Rectangle(\n",
    "            xy=(-0.004, i),\n",
    "            width=0.004,\n",
    "            height=1,\n",
    "            color=color,\n",
    "            lw=0,\n",
    "            transform=ax.get_yaxis_transform(),\n",
    "            clip_on=False,\n",
    "        )\n",
    "    )\n",
    "\n",
    "ax.tick_params(\n",
    "    axis=\"x\", which=\"major\", pad=12, length=0\n",
    ")  # extra padding to leave room for the row colors\n",
    "for i, color in enumerate(col_colors):\n",
    "    ax.add_patch(\n",
    "        plt.Rectangle(\n",
    "            xy=(i, -0.02),\n",
    "            width=1,\n",
    "            height=0.02,\n",
    "            color=color,\n",
    "            lw=0,\n",
    "            transform=ax.get_xaxis_transform(),\n",
    "            clip_on=False,\n",
    "        )\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bulk Sensitivity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g, axs = plt.subplots(2, 3, figsize=(18, 12), sharey=\"row\", sharex=\"col\")\n",
    "plt.subplots_adjust(hspace=0.1, wspace=0.1)\n",
    "thresholds = [10, 50, 100, 500]\n",
    "\n",
    "sens = defaultdict(list)\n",
    "for d, df in bulk_cov.query(\"donor_AC > 0\").groupby(\"cell_id\"):\n",
    "    sns.ecdfplot(\n",
    "        df,\n",
    "        x=\"rpm\",\n",
    "        alpha=0.5,\n",
    "        c=cols[0],\n",
    "        log_scale=True,\n",
    "        stat=\"count\",\n",
    "        complementary=True,\n",
    "        ax=axs[0, 0],\n",
    "    )\n",
    "    sns.ecdfplot(\n",
    "        df,\n",
    "        x=\"rpm\",\n",
    "        alpha=0.5,\n",
    "        c=cols[0],\n",
    "        log_scale=True,\n",
    "        complementary=True,\n",
    "        ax=axs[1, 0],\n",
    "    )\n",
    "    for t in thresholds:\n",
    "        sens[t].append(df.query(f\"rpm > {t}\").shape[0] / df.shape[0])\n",
    "for t, s in sens.items():\n",
    "    axs[1, 0].axvline(t, color=\"black\", linestyle=\"--\")\n",
    "    axs[1, 0].text(t, 0.5, f\"{np.mean(s):.2f}\")\n",
    "axs[0, 0].set_title(\"# KNRGLs\")\n",
    "axs[1, 0].set_title(\"% KNRGLs (sensitivity)\")\n",
    "\n",
    "sens = defaultdict(list)\n",
    "for d, df in bulk_cov.query(\"donor_AC == 1\").groupby(\"cell_id\"):\n",
    "    sns.ecdfplot(\n",
    "        df,\n",
    "        x=\"rpm\",\n",
    "        alpha=0.5,\n",
    "        c=cols[0],\n",
    "        log_scale=True,\n",
    "        stat=\"count\",\n",
    "        complementary=True,\n",
    "        ax=axs[0, 1],\n",
    "    )\n",
    "    sns.ecdfplot(\n",
    "        df,\n",
    "        x=\"rpm\",\n",
    "        alpha=0.5,\n",
    "        c=cols[0],\n",
    "        log_scale=True,\n",
    "        complementary=True,\n",
    "        ax=axs[1, 1],\n",
    "    )\n",
    "    for t in thresholds:\n",
    "        sens[t].append(df.query(f\"rpm > {t}\").shape[0] / df.shape[0])\n",
    "for t, s in sens.items():\n",
    "    axs[1, 1].axvline(t, color=\"black\", linestyle=\"--\")\n",
    "    axs[1, 1].text(t, 0.5, f\"{np.mean(s):.2f}\")\n",
    "axs[0, 1].set_title(\"# het KNRGLs\")\n",
    "axs[1, 1].set_title(\"% het KNRGLs (sensitivity)\")\n",
    "\n",
    "sens = defaultdict(list)\n",
    "for d, df in bulk_cov.query(\"donor_AC == 1 and AC > 1\").groupby(\"cell_id\"):\n",
    "    sns.ecdfplot(\n",
    "        df,\n",
    "        x=\"rpm\",\n",
    "        alpha=0.5,\n",
    "        c=cols[0],\n",
    "        log_scale=True,\n",
    "        stat=\"count\",\n",
    "        complementary=True,\n",
    "        ax=axs[0, 2],\n",
    "    )\n",
    "    sns.ecdfplot(\n",
    "        df,\n",
    "        x=\"rpm\",\n",
    "        alpha=0.5,\n",
    "        c=cols[0],\n",
    "        log_scale=True,\n",
    "        complementary=True,\n",
    "        ax=axs[1, 2],\n",
    "    )\n",
    "    for t in thresholds:\n",
    "        sens[t].append(df.query(f\"rpm > {t}\").shape[0] / df.shape[0])\n",
    "for t, s in sens.items():\n",
    "    axs[1, 2].axvline(t, color=\"black\", linestyle=\"--\")\n",
    "    axs[1, 2].text(t, 0.5, f\"{np.mean(s):.2f}\")\n",
    "axs[0, 2].set_title(\"# het KNRGLs in > 1 donor\")\n",
    "axs[1, 2].set_title(\"% het KNRGLs in > 1 donor (sensitivity)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Builk Contamination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g, ax1 = plt.subplots(1, 1, figsize=(6, 6))\n",
    "\n",
    "for d, df in bulk_cov.query(\"donor_AC == 0\").groupby(\"donor_id\"):\n",
    "    sns.ecdfplot(\n",
    "        df,\n",
    "        x=\"rpm\",\n",
    "        alpha=0.5,\n",
    "        c=cols[0],\n",
    "        log_scale=True,\n",
    "        stat=\"count\",\n",
    "        complementary=True,\n",
    "        ax=ax1,\n",
    "    )\n",
    "ax1.set_title(\"# KNRGLs in other donors\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bulk ROC curve"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g, (ax1, ax2, ax3) = plt.subplots(1, 3, figsize=(18, 6))\n",
    "auc = {}\n",
    "for d, df in bulk_cov.groupby(\"donor_id\"):\n",
    "    fpr, tpr, threshold = roc_curve(df[\"label\"], df[\"rpm\"])\n",
    "    sns.lineplot(x=threshold, y=tpr, ax=ax1, alpha=0.5, c=cols[0], label=\"TPR\")\n",
    "    sns.lineplot(x=threshold, y=fpr, ax=ax1, alpha=0.5, c=cols[1], label=\"FPR\")\n",
    "    sns.lineplot(x=fpr, y=tpr, ax=ax2, alpha=0.5, c=cols[0])\n",
    "    auc[d] = roc_auc_score(df[\"label\"], df[\"rpm\"])\n",
    "\n",
    "# remove legend duplicates\n",
    "handles, labels = ax1.get_legend_handles_labels()\n",
    "by_label = dict(zip(labels, handles))\n",
    "ax1.legend(by_label.values(), by_label.keys())\n",
    "\n",
    "ax1.set_title(\"TPR/FPR vs RPM threshold\")\n",
    "ax1.set_xlabel(\"RPM threshold\")\n",
    "ax1.set_ylabel(\"TPR/FPR\")\n",
    "ax1.set_xscale(\"log\")\n",
    "ax2.set_title(\"ROC curve\")\n",
    "ax2.set_xlabel(\"False positive rate\")\n",
    "ax2.set_ylabel(\"True positive rate\")\n",
    "sns.barplot(x=auc.values(), y=auc.keys(), ax=ax3)\n",
    "ax3.set_title(\"AUC ROC\")\n",
    "ax3.set_xlabel(\"AUC\")\n",
    "ax3.set_ylabel(\"Donor ID\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bulk Sensitivity in 90x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "meta90x = pd.read_csv(\"/iblm/netapp/data3/mcuoco/sz_slavseq/config/U01_LIBD_wgs.tsv\", sep=\"\\t\")\n",
    "indvs = meta90x.set_index(\"ID\")[[\"WGS_90x_HIPPO\",\"WGS_90x_DLPFC\"]].to_dict('index')\n",
    "ids = meta\n",
    "bulk_cov90x = []\n",
    "for i in indvs:\n",
    "\tdid = meta.set_index(\"libd_id\")[\"donor_id\"].loc[i]\n",
    "\t# check if the files exist\n",
    "\twgs = Path(\"/iblm/netapp/data3/mcuoco/sz_slavseq/resources/chm13v2.0.XY/wgs_calls/90x\") / indvs[i][\"WGS_90x_DLPFC\"] / \"MEI_final_gaussian_genotyped.bed\"\n",
    "\tbam = next(Path(f\"/iblm/netapp/data3/mcuoco/sz_slavseq/results/chm13v2.0.XY/filtered/align/{did}\").rglob(\"gDNA*tagged.sorted.bam\"))\n",
    "\tassert bam.exists(), f\"{bam} does not exist\"\n",
    "\t# get one finished individual\n",
    "\tif Path(wgs).exists():\n",
    "\t\t# read the bed file and filter for L1HS\n",
    "\t\tmdf = (\n",
    "\t\t\tread_megane_bed(wgs)\n",
    "\t\t\t.query(\"family == 'LINE/L1'\")\n",
    "\t\t\t.query(\"subfamily.str.contains('L1HS')\")\n",
    "\t\t\t.query(\"confidence.str.contains('PASS')\")\n",
    "\t\t\t.reset_index(drop=True)\n",
    "\t\t)\n",
    "\t\t# save to temp file and perform bash operations\n",
    "\t\twith NamedTemporaryFile() as tmp:\n",
    "\t\t\tpr.PyRanges(mdf).extend(100).to_bed(tmp.name)\n",
    "\t\t\tlibsize = !samtools view -F 1412 -c {bam}\n",
    "\t\t\tcdf = !samtools view -F 1412 -b {bam} -q 30 | bedtools coverage -a {tmp.name} -b stdin -counts\n",
    "\n",
    "\t\t# wrangle\n",
    "\t\tcdf = pd.DataFrame([r.split(\"\\t\") for r in cdf])[[0,1,2,13,14]]\n",
    "\t\tcdf.columns = [\"Chromosome\",\"Start\",\"End\",\"filter\",\"reads\"]\n",
    "\t\tcdf[\"AC\"] = cdf[\"filter\"].str.split(\";\").apply(lambda x: x[0]).astype(int)\n",
    "\t\tcdf[\"locus\"] = cdf[\"Chromosome\"].astype(str) + \":\" + cdf[\"Start\"].astype(str) + \"-\" + cdf[\"End\"].astype(str)\n",
    "\t\tcdf[\"reads\"] = cdf[\"reads\"].astype(int)\n",
    "\t\tlibsize = int(libsize[0]) \n",
    "\t\tcdf[\"rpm\"] =  cdf[\"reads\"] / libsize * 1e6\n",
    "\t\tcdf[\"donor_id\"] = did\n",
    "\t\tbulk_cov90x.append(cdf)\n",
    "\n",
    "bulk_cov90x = pd.concat(bulk_cov90x).reset_index(drop=True)\n",
    "\n",
    "g, axs = plt.subplots(2, 2, figsize=(12, 12), sharey=\"row\", sharex=\"col\")\n",
    "plt.subplots_adjust(hspace=0.1, wspace=0.1)\n",
    "thresholds = [10, 50, 100, 500]\n",
    "\n",
    "sens = defaultdict(list)\n",
    "for d, df in bulk_cov90x.groupby(\"donor_id\"):\n",
    "\tsns.ecdfplot(df, x=\"rpm\", alpha=0.5, c=cols[0], log_scale=True, stat=\"count\", complementary=True, ax=axs[0,0])\n",
    "\tsns.ecdfplot(df, x=\"rpm\", alpha=0.5, c=cols[0], log_scale=True, complementary=True, ax=axs[1,0])\n",
    "\tfor t in thresholds:\n",
    "\t\tsens[t].append(df.query(f\"rpm > {t}\").shape[0] / df.shape[0])\n",
    "for t, s in sens.items():\n",
    "\taxs[1,0].axvline(t, color=\"black\", linestyle=\"--\")\n",
    "\taxs[1,0].text(t, 0.5, f\"{np.mean(s):.2f}\")\n",
    "axs[0,0].set_title(\"# KNRGLs\")\n",
    "axs[1,0].set_title(\"% KNRGLs (sensitivity)\")\n",
    "\n",
    "sens = defaultdict(list)\n",
    "for d, df in bulk_cov90x.query(\"AC == 1\").groupby(\"donor_id\"):\n",
    "\tsns.ecdfplot(df, x=\"rpm\", alpha=0.5, c=cols[0], log_scale=True, stat=\"count\", complementary=True, ax=axs[0,1])\n",
    "\tsns.ecdfplot(df, x=\"rpm\", alpha=0.5, c=cols[0], log_scale=True, complementary=True, ax=axs[1,1])\n",
    "\tfor t in thresholds:\n",
    "\t\tsens[t].append(df.query(f\"rpm > {t}\").shape[0] / df.shape[0])\n",
    "for t, s in sens.items():\n",
    "\taxs[1,1].axvline(t, color=\"black\", linestyle=\"--\")\n",
    "\taxs[1,1].text(t, 0.5, f\"{np.mean(s):.2f}\")\n",
    "axs[0,1].set_title(\"# het KNRGLs\")\n",
    "axs[1,1].set_title(\"% het KNRGLs (sensitivity)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell Sensitivity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g, axs = plt.subplots(6, 7, figsize=(35, 30), sharex=True)\n",
    "axs = axs.flatten()\n",
    "\n",
    "# make dict of axes\n",
    "axs_dict = {}\n",
    "for i, d in enumerate(meta.donor_id.unique()):\n",
    "    axs_dict[d] = axs[i]\n",
    "    axs[i].set_title(d)\n",
    "    axs[i].set_xlabel(\"reads\")\n",
    "    axs[i].set_ylabel(\"% KNRGLs\")\n",
    "\n",
    "for (c, d), df in tqdm(cell_cov.query(\"donor_AC > 0\").groupby([\"cell_id\", \"donor_id\"])):\n",
    "    sns.ecdfplot(\n",
    "        df,\n",
    "        x=\"reads\",\n",
    "        alpha=0.5,\n",
    "        c=cols[0],\n",
    "        log_scale=True,\n",
    "        complementary=True,\n",
    "        ax=axs_dict[d],\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g, axs = plt.subplots(6, 7, figsize=(35, 30), sharex=True)\n",
    "axs = axs.flatten()\n",
    "\n",
    "# make dict of axes\n",
    "axs_dict = {}\n",
    "for i, d in enumerate(meta.donor_id.unique()):\n",
    "    axs_dict[d] = axs[i]\n",
    "    axs[i].set_title(d)\n",
    "    axs[i].set_xlabel(\"reads\")\n",
    "    axs[i].set_ylabel(\"# KNRGLs\")\n",
    "    axs[i].set_ylim(0, 50)\n",
    "\n",
    "for (c, d), df in tqdm(\n",
    "    cell_cov.query(\"donor_AC == 0\").groupby([\"cell_id\", \"donor_id\"])\n",
    "):\n",
    "    sns.ecdfplot(\n",
    "        df,\n",
    "        x=\"reads\",\n",
    "        alpha=0.5,\n",
    "        c=cols[0],\n",
    "        log_scale=True,\n",
    "        stat=\"count\",\n",
    "        complementary=True,\n",
    "        ax=axs_dict[d],\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Single Cell ROC curve"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g, axs = plt.subplots(6, 7, figsize=(35, 30), sharex=True)\n",
    "axs = axs.flatten()\n",
    "\n",
    "# make dict of axes\n",
    "axs_dict = {}\n",
    "for i, d in enumerate(meta.donor_id.unique()):\n",
    "    axs_dict[d] = axs[i]\n",
    "    axs[i].set_title(d)\n",
    "\n",
    "auc = {}\n",
    "for (c, d), df in tqdm(cell_cov.groupby([\"cell_id\", \"donor_id\"])):\n",
    "    fpr, tpr, threshold = roc_curve(df[\"label\"], df[\"rpm\"])\n",
    "    sns.lineplot(x=fpr, y=tpr, ax=axs_dict[d], alpha=0.5, c=cols[0])\n",
    "    auc[(c, d)] = roc_auc_score(df[\"label\"], df[\"rpm\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## cell coverage by donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = (\n",
    "    cell_cov[(cell_cov.donor_AC == 1) & (cell_cov.donor_id != \"CommonBrain\")]\n",
    "    .groupby([\"locus\", \"cell_id\", \"donor_id\", \"tissue\"])[\"reads\"]\n",
    "    .sum()\n",
    "    .reset_index(name=\"reads\")\n",
    ")\n",
    "df[\"donor_tissue\"] = df[\"donor_id\"].str.cat(df[\"tissue\"], sep=\"_\")\n",
    "\n",
    "g = sns.ecdfplot(\n",
    "    df, x=\"reads\", hue=\"donor_tissue\", alpha=0.5, complementary=True, stat=\"count\"\n",
    ")\n",
    "g.set_xscale(\"log\")\n",
    "g.legend_.remove()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# g, ax = plt.subplots(1, 1, figsize=(10, 10))\n",
    "for t in [3, 5, 10, 20, 50, 100]:\n",
    "    donor_total = df.groupby([\"donor_id\"])[\"cell_id\"].nunique()\n",
    "    n_cov_donor = df.groupby([\"locus\", \"donor_id\"])[\"reads\"].apply(lambda x: sum(x > t))\n",
    "    # frac_cov_donor = n_cov_donor / donor_total\n",
    "    donor_tissue_total = df.groupby([\"donor_id\", \"tissue\"])[\"cell_id\"].nunique()\n",
    "    n_cov_tissue = (\n",
    "        df.groupby([\"locus\", \"donor_id\", \"tissue\"])[\"reads\"]\n",
    "        .apply(lambda x: sum(x > t))\n",
    "        .reset_index(name=\"n_cells\")\n",
    "    )\n",
    "    break\n",
    "\n",
    "sns.displot(\n",
    "    n_cov_tissue,\n",
    "    x=\"n_cells\",\n",
    "    col=\"tissue\",\n",
    "    hue=\"donor_id\",\n",
    "    kind=\"ecdf\",\n",
    "    complementary=True,\n",
    "    stat=\"count\",\n",
    "    hue_order=[\"1\", \"10\", \"11\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.jointplot(\n",
    "    n_cov_tissue,\n",
    "    x=\"DLPFC\",\n",
    "    y=\"HIP\",\n",
    "    kind=\"hist\",\n",
    "    joint_kws={\"discrete\": True},\n",
    "    marginal_kws={\"discrete\": True},\n",
    "    marginal_ticks=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.jointplot(\n",
    "    n_cov_tissue[~((n_cov_tissue.DLPFC == 0) & (n_cov_tissue.HIP == 0))],\n",
    "    x=\"DLPFC\",\n",
    "    y=\"HIP\",\n",
    "    kind=\"hist\",\n",
    "    joint_kws={\"discrete\": True},\n",
    "    marginal_kws={\"discrete\": True},\n",
    "    marginal_ticks=True,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
