{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b47ead2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import pyarrow.parquet as pq\n",
    "import warnings, math\n",
    "\n",
    "warnings.filterwarnings(\"ignore\", category=FutureWarning)\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "import pandas as pd\n",
    "\n",
    "tqdm.pandas()\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pyranges as pr\n",
    "from pyslavseq.preprocessing import collate_labels\n",
    "from pyslavseq.plotting import datashader_plot\n",
    "\n",
    "HUE_ORDER = [\"KNRGL\", \"OTHER\", \"KRGL\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d90dfb2f",
   "metadata": {},
   "source": [
    "# TODO\n",
    "\n",
    "- mappability\n",
    "- "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1411ed93",
   "metadata": {},
   "source": [
    "## Load data\n",
    "\n",
    "1. meta data\n",
    "1. bulk peaks\n",
    "1. single-cell peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c5fb262",
   "metadata": {},
   "outputs": [],
   "source": [
    "# meta data\n",
    "meta = pd.read_csv(snakemake.config[\"samples\"], sep=\"\\t\", dtype={\"sample_id\": str, \"tissue_id\": str, \"donor_id\": str})  # type: ignore\n",
    "donors = pd.read_csv(snakemake.config[\"donors\"], sep=\"\\t\", dtype={\"donor_id\": str})  # type: ignore\n",
    "meta = meta.merge(donors, on=[\"donor_id\"]).rename(columns={\"sample_id\": \"cell_id\"})\n",
    "\n",
    "# bulk peaks\n",
    "bulk = pd.read_parquet(snakemake.input.bulk_peaks)  # type: ignore\n",
    "\n",
    "# single-cell peaks\n",
    "data = []\n",
    "for f in tqdm(snakemake.input.cell_peaks):  # type: ignore\n",
    "    df = pq.read_table(f).to_pandas()\n",
    "    cell_id = Path(f).name.rstrip(\"_labelled.pqt\")\n",
    "    df[\"cell_id\"] = cell_id\n",
    "    df[\"tissue\"] = \"DLPFC\" if \"usd\" in cell_id.lower() else \"HIP\"\n",
    "    data.append(df)\n",
    "\n",
    "data = pd.concat(data).reset_index(drop=True).merge(meta, on=\"cell_id\")\n",
    "\n",
    "\n",
    "ndonors = data[\"donor_id\"].nunique()\n",
    "ncells = data[\"cell_id\"].nunique()\n",
    "print(f\"Loaded {len(data)} peaks from {ncells} cells from {ndonors} donors.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69862d17",
   "metadata": {},
   "source": [
    "## Preprocess data\n",
    "\n",
    "1. Remove peaks without any reads with >= 60 MAPQ \n",
    "2. Compute sharing of peaks across cells and donors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99a119b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove peaks without any reads with at least a 30 MAPQ score\n",
    "n_peaks = len(data)\n",
    "data = data.query(\"max_mapq >= 60\").reset_index(drop=True)\n",
    "print(f\"Removed {n_peaks - len(data)}/{n_peaks} peaks with MAX MAPQ < 60\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cefa5d1",
   "metadata": {},
   "source": [
    "Cluster peak across cells and see how big the clusters are\n",
    "\n",
    "TODO: inspect very large clusters (will they be gone with new mapq read filter?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dda3ac7",
   "metadata": {},
   "outputs": [],
   "source": [
    "merged = pr.PyRanges(data).merge().df\n",
    "merged[\"width\"] = merged[\"End\"] - merged[\"Start\"]\n",
    "\n",
    "# plot width of clusters\n",
    "g, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 6), sharex=True, sharey=False)\n",
    "sns.ecdfplot(data, x=\"width\", stat=\"count\", ax=ax1, log_scale=True)\n",
    "ax1.set_title(\"Single-cell peaks\")\n",
    "sns.ecdfplot(merged, x=\"width\", stat=\"count\", ax=ax2, log_scale=True)\n",
    "ax2.set_title(\"Peak clusters\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f233661",
   "metadata": {},
   "outputs": [],
   "source": [
    "def donor_cells(df):\n",
    "    \"\"\"\n",
    "    Count how many cells each peaks is present in for each donor.\n",
    "    \"\"\"\n",
    "    assert \"Cluster\" in df.columns, \"Cluster column missing.\"\n",
    "    assert \"tissue\" in df.columns, \"Tissue column missing.\"\n",
    "    assert df[\"donor_id\"].nunique() == 1, \"Multiple donors present.\"\n",
    "    assert \"cell_id\" in df.columns, \"Cell ID column missing.\"\n",
    "\n",
    "    cells_per_peak = (\n",
    "        df.groupby([\"Cluster\", \"tissue\"], observed=True)[\"cell_id\"]\n",
    "        .nunique()\n",
    "        .to_frame()\n",
    "        .reset_index()\n",
    "        .pivot(index=\"Cluster\", columns=\"tissue\", values=\"cell_id\")\n",
    "        .fillna(0)\n",
    "        .astype(int)\n",
    "    )\n",
    "\n",
    "    cells_per_peak = cells_per_peak.rename(\n",
    "        columns={\"DLPFC\": \"n_DLPFC_cells\", \"HIP\": \"n_HIP_cells\"}\n",
    "    )\n",
    "    cells_per_peak[\"n_donor_cells\"] = (\n",
    "        cells_per_peak[\"n_DLPFC_cells\"] + cells_per_peak[\"n_HIP_cells\"]\n",
    "    )\n",
    "\n",
    "    return df.merge(cells_per_peak, on=\"Cluster\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "568d49a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute sharing of peaks across cells and donors\n",
    "data = pr.PyRanges(data).cluster().df\n",
    "data[\"Cluster\"] = data[\"Cluster\"].astype(\"category\")\n",
    "data = data.groupby(\"donor_id\").apply(donor_cells).reset_index(drop=True)\n",
    "cells_per_peak = data.groupby(\"Cluster\", observed=True)[\"cell_id\"].nunique()\n",
    "donors_per_peak = data.groupby(\"Cluster\", observed=True)[\"donor_id\"].nunique()\n",
    "data[\"n_cells\"] = data[\"Cluster\"].map(cells_per_peak)\n",
    "data[\"n_donors\"] = data[\"Cluster\"].map(donors_per_peak)\n",
    "data[\"cells_per_donor\"] = data[\"n_cells\"] / data[\"n_donors\"]\n",
    "data[\"cells_per_donor\"] = data[\"cells_per_donor\"].apply(\n",
    "    lambda x: 0 if x == math.inf else x\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b91e38a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute sharing across bulk data\n",
    "# TODO: something is going wrong here, resulting in too many KNRGLs\n",
    "merged = (\n",
    "    pr.PyRanges(merged)\n",
    "    .count_overlaps(pr.PyRanges(bulk), overlap_col=\"n_bulk_donors\")\n",
    "    .df\n",
    ")\n",
    "merged.index = merged.index + 1\n",
    "data[\"n_bulk_donors\"] = data[\"Cluster\"].map(merged[\"n_bulk_donors\"])\n",
    "data[\"n_bulk_donors\"] = data[\"n_bulk_donors\"].apply(lambda x: 37 if x > 37 else x)\n",
    "data[\"cells_per_bulk_donor\"] = data[\"n_cells\"] / data[\"n_bulk_donors\"]\n",
    "data[\"cells_per_bulk_donor\"] = data[\"cells_per_bulk_donor\"].apply(\n",
    "    lambda x: 0 if x == math.inf else x\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "890bd42e",
   "metadata": {},
   "source": [
    "## Display statistics of unlabeled peaks\n",
    "\n",
    "1. number of peaks per cell per donor, TODO: check for correlations with clinical features and number of reads\n",
    "1. width\n",
    "1. n_reads\n",
    "1. rpm\n",
    "1. n_ref_reads\n",
    "1. n_unique_5end\n",
    "1. frac_unique_5end\n",
    "1. n_unique_3end\n",
    "1. frac_unique_5end\n",
    "1. n_duplicates\n",
    "1. frac_duplicates\n",
    "1. n_contigs\n",
    "1. frac_contigs\n",
    "1. n_proper_pairs\n",
    "1. frac_proper_pairs\n",
    "1. num_supp_alignments_mean\n",
    "1. frac_mean_supp_alignments\n",
    "1. mate_alignment_score_normed_mean\n",
    "1. min_mapq\n",
    "1. n_cells\n",
    "1. n_donors\n",
    "1. n_bulk_donors\n",
    "1. cells_per_donor\n",
    "1. cells_per_bulk_donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ffb7e8f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = (\n",
    "    data.groupby([\"libd_id\", \"race\", \"diagnosis\", \"donor_id\", \"cell_id\", \"age\"])\n",
    "    .size()\n",
    "    .reset_index(name=\"n_peaks\")\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(1, 3, figsize=(20, 6))\n",
    "df.sort_values(\"diagnosis\", ascending=False, inplace=True)\n",
    "sns.boxplot(data=df, y=\"libd_id\", x=\"n_peaks\", ax=ax[0])\n",
    "mean_peaks = data.groupby(\"cell_id\", observed=True).size().mean()\n",
    "\n",
    "# annotate ax with mean peaks\n",
    "ax[0].axvline(mean_peaks, color=\"red\", linestyle=\"--\")\n",
    "\n",
    "for _, df in data.groupby(\"cell_id\"):\n",
    "    sns.ecdfplot(\n",
    "        df,\n",
    "        x=\"n_reads\",\n",
    "        c=sns.color_palette()[0],\n",
    "        alpha=0.2,\n",
    "        stat=\"count\",\n",
    "        ax=ax[1],\n",
    "        log_scale=True,\n",
    "    )\n",
    "    sns.ecdfplot(\n",
    "        df, x=\"width\", c=sns.color_palette()[0], alpha=0.2, stat=\"count\", ax=ax[2]\n",
    "    )\n",
    "ax[1].set_ylabel(\"# peaks\")\n",
    "ax[2].set_ylabel(\"# peaks\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0faef44e",
   "metadata": {},
   "outputs": [],
   "source": [
    "features_scale = [\n",
    "    (\"width\", False),\n",
    "    (\"n_reads\", True),\n",
    "    (\"n_ref_reads\", True),\n",
    "    (\"n_unique_5end\", True),\n",
    "    (\"n_unique_3end\", True),\n",
    "    (\"n_duplicates\", True),\n",
    "    (\"n_contigs\", True),\n",
    "    (\"n_proper_pairs\", True),\n",
    "    (\"num_supp_alignments_mean\", True),\n",
    "    (\"mate_alignment_score_normed_mean\", False),\n",
    "    (\"alignment_score_normed_mean\", False),\n",
    "    (\"orientation_bias\", False),\n",
    "    (\"min_mapq\", False),\n",
    "    (\"n_cells\", False),\n",
    "    (\"n_donors\", False),\n",
    "    (\"n_bulk_donors\", False),\n",
    "    (\"n_HIP_cells\", False),\n",
    "    (\"n_DLPFC_cells\", False),\n",
    "    (\"n_donor_cells\", False),\n",
    "]\n",
    "features, scales = zip(*features_scale)\n",
    "n_plots = len(features_scale)\n",
    "nrows = math.floor(math.sqrt(n_plots))\n",
    "ncols = math.ceil(n_plots / nrows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31058aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "g, axs = plt.subplots(nrows, ncols, figsize=(ncols * 6, nrows * 6))\n",
    "\n",
    "# setup colors\n",
    "colors = sns.color_palette(\"tab10\", n_colors=2)\n",
    "opts = {\n",
    "    \"hue_order\": [\"CAUC\", \"AA\"],\n",
    "    \"hue\": \"race\",\n",
    "    \"palette\": {\"CAUC\": colors[0], \"AA\": colors[1]},\n",
    "    \"alpha\": 0.5,\n",
    "}\n",
    "\n",
    "for ax, (f, s) in zip(axs.flatten(), features_scale):\n",
    "    for d, df in data.groupby(\"libd_id\"):\n",
    "        sns.ecdfplot(data=df, x=f, log_scale=s, ax=ax, **opts)\n",
    "    ax.get_legend().remove()\n",
    "\n",
    "g.legend(\n",
    "    title=\"Race\",\n",
    "    labels=data[\"race\"].unique(),\n",
    "    loc=\"center right\",\n",
    "    bbox_to_anchor=(1, 0.5),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "896549fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# by chromosome\n",
    "df = (\n",
    "    data.groupby([\"Chromosome\", \"race\", \"diagnosis\", \"cell_id\", \"age\"])\n",
    "    .size()\n",
    "    .reset_index(name=\"n_peaks\")\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(1, 3, figsize=(21, 6))\n",
    "df.sort_values(\"n_peaks\", inplace=True)\n",
    "sns.boxplot(data=df, y=\"Chromosome\", x=\"n_peaks\", ax=ax[0])\n",
    "sns.ecdfplot(data, x=\"n_reads\", hue=\"Chromosome\", alpha=0.5, ax=ax[1], log_scale=True)\n",
    "sns.ecdfplot(data, x=\"width\", hue=\"Chromosome\", alpha=0.5, ax=ax[2])\n",
    "ax[1].set_ylabel(\"# peaks\")\n",
    "ax[1].get_legend().remove()\n",
    "ax[2].set_ylabel(\"# peaks\")\n",
    "\n",
    "# move legend to the right\n",
    "ax[2].legend(\n",
    "    title=\"Chromosome\",\n",
    "    labels=data[\"Chromosome\"].unique(),\n",
    "    loc=\"center left\",\n",
    "    bbox_to_anchor=(1, 0.5),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09be3448",
   "metadata": {},
   "outputs": [],
   "source": [
    "g, axs = plt.subplots(nrows, ncols, figsize=(ncols * 6, nrows * 6), sharey=True)\n",
    "\n",
    "for ax, (f, s) in zip(axs.flatten(), features_scale):\n",
    "    sns.ecdfplot(data=data, x=f, log_scale=s, ax=ax, hue=\"Chromosome\", alpha=0.5)\n",
    "    ax.get_legend().remove()\n",
    "\n",
    "g.legend(\n",
    "    title=\"Chromosome\",\n",
    "    labels=data[\"Chromosome\"].unique(),\n",
    "    loc=\"center left\",\n",
    "    bbox_to_anchor=(1, 0.5),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc2030ea",
   "metadata": {},
   "source": [
    "## 4. Collate labels of the peaks\n",
    "\n",
    "a. Find additional KRGLs by looking for overlaps with bulk SLAVseq\n",
    "b. Find additional KRGLs by looking for overlaps with single-cell SLAVseq\n",
    "c. Find additional KNRGLs by looking for overlaps with bulk SLAVseq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9597ae4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[\"label\"] = data.progress_apply(collate_labels, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42006f96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get germline labels for each donor\n",
    "# TODO: use bulk here or single-cell data as reference?\n",
    "def germline_distance(donor_df: pd.DataFrame) -> pd.DataFrame:\n",
    "    assert len(donor_df[\"donor_id\"].unique()) == 1, \"Expected only one donor\"\n",
    "    d = donor_df[\"donor_id\"].unique()[0]\n",
    "    gdf = pr.PyRanges(bulk.query('donor_id == @d and label != \"OTHER\"')).merge()\n",
    "    donor_df = pr.PyRanges(donor_df).df\n",
    "    donor_df[\"germline_dist\"] = (\n",
    "        pr.PyRanges(donor_df).nearest(gdf, overlap=False).df[\"Distance\"].abs()\n",
    "    )\n",
    "    return donor_df\n",
    "\n",
    "\n",
    "data = (\n",
    "    data.groupby([\"donor_id\"])\n",
    "    .progress_apply(germline_distance)\n",
    "    .reset_index(drop=True)\n",
    "    .sort_values([\"Chromosome\", \"Start\"])\n",
    ")\n",
    "\n",
    "features_scale.append((\"germline_dist\", True))\n",
    "features, scale = zip(*features_scale)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b5696f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = [\n",
    "    \"primer_sites\",\n",
    "    \"l1hs\",\n",
    "    \"l1pa2\",\n",
    "    \"l1pa3\",\n",
    "    \"l1pa4\",\n",
    "    \"l1pa5\",\n",
    "    \"l1pa6\",\n",
    "    \"megane_gaussian\",\n",
    "    \"megane_breakpoints\",\n",
    "    \"graffite\",\n",
    "    \"xtea\",\n",
    "    \"bulk\",\n",
    "]\n",
    "data[\"other\"] = data[labels].sum(axis=1) == 0\n",
    "labels.append(\"other\")\n",
    "\n",
    "g, axs = plt.subplots(nrows, ncols, figsize=(ncols * 6, nrows * 6), sharey=True)\n",
    "axs = axs.flatten()\n",
    "\n",
    "for ax, (f, s) in zip(axs[0 : len(features)], features_scale):\n",
    "    for l in labels:\n",
    "        sns.ecdfplot(data=data[data[l]], x=f, log_scale=s, ax=ax)\n",
    "\n",
    "g.legend(title=\"annotation\", labels=labels, loc=\"center left\", bbox_to_anchor=(1, 0.5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7f4cf63",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = (\n",
    "    data.groupby([\"label\", \"race\", \"diagnosis\", \"cell_id\", \"age\", \"libd_id\"])\n",
    "    .size()\n",
    "    .reset_index(name=\"n_peaks\")\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(1, 3, figsize=(21, 10))\n",
    "df.sort_values(\"n_peaks\", inplace=True)\n",
    "sns.boxplot(\n",
    "    data=df, y=\"libd_id\", x=\"n_peaks\", hue=\"label\", ax=ax[0], hue_order=HUE_ORDER\n",
    ")\n",
    "sns.ecdfplot(\n",
    "    data,\n",
    "    x=\"n_reads\",\n",
    "    hue=\"label\",\n",
    "    alpha=0.5,\n",
    "    ax=ax[1],\n",
    "    log_scale=True,\n",
    "    hue_order=HUE_ORDER,\n",
    ")\n",
    "sns.ecdfplot(data, x=\"width\", hue=\"label\", alpha=0.5, ax=ax[2], hue_order=HUE_ORDER)\n",
    "\n",
    "means = data.groupby([\"cell_id\", \"label\"]).size().groupby(\"label\").mean()\n",
    "print(\n",
    "    f\"\"\"\n",
    "       Mean KRNGL peaks per cell: {means[\"KNRGL\"]:0.2f}\n",
    "       Mean KRGL peaks per cell: {means[\"KRGL\"]:0.2f}\n",
    "       Mean OTHER peaks per cell: {means[\"OTHER\"]:0.2f}\n",
    "       \"\"\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0301784",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = data.melt(\n",
    "    id_vars=[\"label\", \"cell_id\"],\n",
    "    value_vars=features,\n",
    "    var_name=\"feature\",\n",
    "    value_name=\"value\",\n",
    ")\n",
    "\n",
    "g = sns.displot(\n",
    "    data=df,\n",
    "    x=\"value\",\n",
    "    hue=\"label\",\n",
    "    col=\"feature\",\n",
    "    kind=\"ecdf\",\n",
    "    col_wrap=6,\n",
    "    facet_kws={\"sharex\": False},\n",
    ")\n",
    "\n",
    "for ax in g.axes:\n",
    "    for f, s in features_scale:\n",
    "        text = ax.title.get_text().split(\" = \")[-1]\n",
    "        if f == text:\n",
    "            ax.set_xscale(\"log\" if s else \"linear\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed4ff36b",
   "metadata": {},
   "outputs": [],
   "source": [
    "plots = [\n",
    "    (\"germline_dist\", \"n_bulk_donors\", (True, False)),\n",
    "    (\"germline_dist\", \"n_donors\", (True, False)),\n",
    "    (\"germline_dist\", \"n_cells\", (True, False)),\n",
    "    (\"germline_dist\", \"cells_per_donor\", (True, False)),\n",
    "    (\"germline_dist\", \"n_donor_cells\", (True, False)),\n",
    "    (\"germline_dist\", \"n_HIP_cells\", (True, False)),\n",
    "    (\"germline_dist\", \"n_DLPFC_cells\", (True, False)),\n",
    "    (\"n_cells\", \"n_donors\", (False, False)),\n",
    "    (\"n_cells\", \"n_bulk_donors\", (False, False)),\n",
    "    (\"n_cells\", \"n_donor_cells\", (False, False)),\n",
    "    (\"n_donors\", \"n_bulk_donors\", (False, False)),\n",
    "]\n",
    "\n",
    "for i, (x, y, s) in enumerate(plots):\n",
    "    datashader_plot(data, x, y, s, plot_width=100, plot_height=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba86cb6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get all peaks that are KRGL in bulk\n",
    "brdf = pr.PyRanges(bulk.query(\"label == 'KRGL'\"))\n",
    "# find overlaps across cells\n",
    "data = pr.PyRanges(data).count_overlaps(brdf, overlap_col=\"bulk_ref\").df\n",
    "# label as bulk_ref if overlap\n",
    "data[\"bulk_ref\"] = data[\"bulk_ref\"] > 0\n",
    "\n",
    "# report new KRGL peaks per donor\n",
    "for d, df in data.groupby(\"donor_id\"):\n",
    "    new_krgls = df.query(\"label != 'KRGL' and bulk_ref\")\n",
    "    krgl_low_germ_dist = new_krgls.query(\"germline_dist < 20000\")\n",
    "    lost_knrgls = df.query(\"label == 'KNRGL' and bulk_ref\")\n",
    "    knrgl_low_germ_dist = lost_knrgls.query(\"germline_dist < 20000\")\n",
    "    print(\n",
    "        f\"\"\"Donor {d}:\n",
    "\t\t\t\t\t{len(new_krgls)} ({len(krgl_low_germ_dist)} germ_dist < 20kb) new KRGLs \n",
    "\t\t\t\t\t{len(lost_knrgls)} ({len(knrgl_low_germ_dist)} germ_dist < 20kb) lost KNRGLs\n",
    "\t\t\t\"\"\"\n",
    "    )\n",
    "\n",
    "data.loc[data[\"bulk_ref\"], \"label\"] = \"KRGL\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10df9064",
   "metadata": {},
   "outputs": [],
   "source": [
    "knrgl_low_germ_dist = data.query(\"label == 'KNRGL' and germline_dist < 20000\")\n",
    "print(f\"Found {len(knrgl_low_germ_dist)} KNRGL peaks with germline distance < 20kb.\")\n",
    "\n",
    "new_krgls = data.query(\"label == 'OTHER' and bulk\")\n",
    "print(f\"Added {len(new_krgls)} new KNRGLs.\")\n",
    "\n",
    "data.loc[(data[\"bulk\"]) & (data[\"label\"] == \"OTHER\"), \"label\"] = \"KNRGL\"\n",
    "knrgl_low_germ_dist = data.query(\"label == 'KNRGL' and germline_dist < 20000\")\n",
    "print(f\"Found {len(knrgl_low_germ_dist)} KNRGL peaks with germline distance < 20kb.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6b08c37",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = (\n",
    "    data.groupby([\"label\", \"race\", \"diagnosis\", \"cell_id\", \"age\", \"libd_id\"])\n",
    "    .size()\n",
    "    .reset_index(name=\"n_peaks\")\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(1, 3, figsize=(21, 10))\n",
    "df.sort_values(\"n_peaks\", inplace=True)\n",
    "sns.boxplot(\n",
    "    data=df, y=\"libd_id\", x=\"n_peaks\", hue=\"label\", ax=ax[0], hue_order=HUE_ORDER\n",
    ")\n",
    "sns.ecdfplot(\n",
    "    data,\n",
    "    x=\"n_reads\",\n",
    "    hue=\"label\",\n",
    "    alpha=0.5,\n",
    "    ax=ax[1],\n",
    "    log_scale=True,\n",
    "    hue_order=HUE_ORDER,\n",
    ")\n",
    "sns.ecdfplot(data, x=\"width\", hue=\"label\", alpha=0.5, ax=ax[2], hue_order=HUE_ORDER)\n",
    "\n",
    "means = data.groupby([\"cell_id\", \"label\"]).size().groupby(\"label\").mean()\n",
    "print(\n",
    "    f\"\"\"\n",
    "       Mean KRNGL peaks per cell: {means[\"KNRGL\"]:0.2f}\n",
    "       Mean KRGL peaks per cell: {means[\"KRGL\"]:0.2f}\n",
    "       Mean OTHER peaks per cell: {means[\"OTHER\"]:0.2f}\n",
    "       \"\"\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8814dc96",
   "metadata": {},
   "source": [
    "## Remove peaks at reference L1 insertions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0cb1d2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove reference insertions and peaks with low mapping quality\n",
    "print(\"Filtering peaks\")\n",
    "print(\"Found {} peaks with reference reads.\".format(len(data[data[\"n_ref_reads\"] > 0])))\n",
    "# print(\"Found {} peaks with reference clusters.\".format(data[\"ref\"].sum()))\n",
    "print(\n",
    "    \"Found {} peaks with bulk SLAVseq reference clusters.\".format(\n",
    "        data[\"bulk_ref\"].sum()\n",
    "    )\n",
    ")\n",
    "print(\"Found {} peaks at primer sites.\".format(data[\"primer_sites\"].sum()))\n",
    "for l in [\"l1hs\", \"l1pa2\", \"l1pa3\", \"l1pa4\", \"l1pa5\", \"l1pa6\"]:\n",
    "    print(\"Found {} peaks at {} sites.\".format(data[l].sum(), l))\n",
    "\n",
    "data = data.query(\"label != 'KRGL'\").reset_index(drop=True)\n",
    "print(\"{} peaks remain after filtering\".format(len(data)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "918d18df",
   "metadata": {},
   "source": [
    "## Visualize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d284715",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, (x, y, s) in enumerate(plots):\n",
    "    datashader_plot(data, x, y, s, plot_width=100, plot_height=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59c77482",
   "metadata": {},
   "source": [
    "## Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4ef5ef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_parquet(snakemake.output[0], index=False)  # type: ignore"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
